arch/risc-v/src/bl808/bl808_wdt.c

```diff
@@ -0,0 +1,448 @@
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  priv->started = true;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_stop
+ *
+ * Description:
+ *   Stop the watchdog.
+ *
+ * Input parameters:
+ *   lower  - A pointer to the lower-half driver of the watchdog.
+ *
+ * Returned Value:
+ *   Error status. Always returns OK.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_stop(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* WDTs are stopped by setting the input clock to NONE.
+   * This is done to allow calling watchdog_stop and then get
+   * the time left to timeout afterwards. If we used the
+   * watchdog enable bits, the counter would reset to 0 when stopped.
+   */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_NONE << WDT_CLKSEL_SHIFT);
+
+  priv->started = false;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_keepalive
+ *
+ * Description:
+ *   Reset the watchdog to keep it running.
+ *
+ * Input parameters:
+ *   lower  - A pointer to the lower-half driver of the watchdog.
+ *
+ * Returned Value:
+ *   Error status. Returns an IO error if the watchdog is not
+ *   running, otherwise returns OK.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_keepalive(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* Check that the watchdog is running */
+
+  if (priv->started == false)
+    {
+      return -EIO;
+    }
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), 0, WDT_CLEAR_COUNT);
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_getstatus
+ *
+ * Description:
+ *   Get current watchdog status. Returns to status parameter.
+ *
+ * Input parameters:
+ *   lower  - A pointer to the lower-half driver of the watchdog.
+ *   status - Return location for the watchdog status.
+ *
+ * Returned Value:
+ *   Error status. Always returns OK.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_getstatus(FAR struct watchdog_lowerhalf_s *lower,
+                          FAR struct watchdog_status_s *status)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  status->flags = priv->started
+    | ((priv->callback == NULL) << 1)
+    | ((priv->callback != NULL) << 2);
```

Comment:
IMHO usage of `WDFLAGS_ACTIVE`, `WDFLAGS_RESET` and `WDFLAGS_CAPTURE` is more transparent